libyul/optimiser/DataFlowAnalyzer.cpp

```diff
@@ -53,33 +100,48 @@ void DataFlowAnalyzer::operator()(VariableDeclaration& _varDecl)
 	m_variableScopes.back().variables += names;
 
 	if (_varDecl.value)
+	{
+		clearKnowledgeIfInvalidated(*_varDecl.value);
 		visit(*_varDecl.value);
+	}
 
 	handleAssignment(names, _varDecl.value.get());
 }
 
 void DataFlowAnalyzer::operator()(If& _if)
 {
+	clearKnowledgeIfInvalidated(*_if.condition);
+	InvertibleMap<YulString, YulString> storage = m_storage;
+
 	ASTModifier::operator()(_if);
 
+	joinKnowledge(storage);
+
 	Assignments assignments;
 	assignments(_if.body);
 	clearValues(assignments.names());
 }
 
 void DataFlowAnalyzer::operator()(Switch& _switch)
 {
+	clearKnowledgeIfInvalidated(*_switch.expression);
 	visit(*_switch.expression);
 	set<YulString> assignedVariables;
 	for (auto& _case: _switch.cases)
 	{
+		InvertibleMap<YulString, YulString> storage = m_storage;
 		(*this)(_case.body);
+		joinKnowledge(storage);
```

Comment:
Why only storage knowledge is joined?

Reply:
Oh wow! This is of course totally wrong!